docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,493 @@
+go modules `foo` and `foo/v2` which could use a types or API go module say
+`foo/api`. This `foo/api` go module would be perpetually on `v1.x` and only
+accept non-breaking changes. This would then allow other modules to be
+compatible with either `foo` or `foo/v2` as long as the inter-module API only
+depends on the types in `foo/api`. It would also allow modules `foo` and `bar`
+to depend on each other in that both of them could depend on `foo/api` and
+`bar/api` without `foo` directly depending on `bar` and vice versa.
+
+This is similar to the naive mitigation described above except that it separates
+the types into separate go modules which in and of itself could be used to
+break circular module dependencies. It has the same problems as the naive solution,
+otherwise, which we could rectify by:
+1. removing all state machine breaking code from the API module (ex. `ValidateBasic` and any other interface methods)
+2. embedding the correct file descriptors for unknown field filtering in the binary
+
+#### Migrate all interface methods on API types to handlers 
+
+To solve 1), we need to remove all interface implementations from generated
+types and instead use a handler approach which essentially means that given
+a type `X`, we have some sort of resolver which allows us to resolve interface
+implementations for that type (ex. `sdk.Msg` or `authz.Authorization`). For
+example:
+
+```go
+func (k Keeper) DoSomething(msg MsgDoSomething) error {
+	var validateBasicHandler ValidateBasicHandler
+	err := k.resolver.Resolve(&validateBasic, msg)
+	if err != nil {
+		return err
+    }   
+	
+	err = validateBasicHandler.ValidateBasic()
+	...
+}
+```
+
+In the case of some methods on `sdk.Msg`, we could replace them with declarative
+annotations. For instance, `GetSigners` can already be replaced by the protobuf
+annotation `cosmos.msg.v1.signer`. In the future, we may consider some sort
+of protobuf validation framework (like https://github.com/bufbuild/protoc-gen-validate
+but more Cosmos-specific) to replace `ValidateBasic`.
+
+#### Pinned FileDescriptor's
+
+To solve 2), state machine modules must be able to specify what the version of
+the protobuf files was that they were built against. For instance if the API
+module for `foo` upgrades to `foo/v2`, the original `foo` module still needs
+a copy of the original protobuf files it was built with so that ADR 020
+unknown field filtering will reject `MsgDoSomething` when `condition` is
+set.
+
+The simplest way to do this may be to embed the protobuf `FileDescriptor`s into
+the module itself so that these `FileDescriptor`s are used at runtime rather 
+than the ones that are built into the `foo/api` which may be different. Using
+[buf build](https://docs.buf.build/build/usage#output-format), [go embed](https://pkg.go.dev/embed),
+and a build script we can probably come up with a solution for embedding
+`FileDescriptor`s into modules that is fairly straightforward.
```

Comment:
don't we need to provide that information in the seriazlied bytes as well?

Reply:
Which serialized bytes?

Reply:
In Msg bytes.